# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/TextLeafRange.cpp
# Commit: 16aee29be1b7
# Full Hash: 16aee29be1b7ae71eae277b97813fe6bf851ac0f
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-27 15:58:57
# Regressor Bug: 1737919
# File Overlap Count: 1
# Description:
#   Bug 1737919 part 1: TextLeafPoint::FindTextAttrsStart: Don't allow the caller to supply origin attributes. r=morgan
#   
#   This was done for LocalAccessibles to avoid the need to fetch attributes twice when we're simultaneously fetching boundaries and returning the attributes to a client.
#   However, GetTextAttributes will soon handle spelling errors, so it will no longer return the same data as GetTextAttributesLocalAcc (which doesn't handle spelling errors).
#   This breaks the equality checks in FindTextAttrsStart.
# ==============================================================================

diff -r 27f0da23c937 -r 16aee29be1b7 accessible/base/TextLeafRange.cpp
--- a/accessible/base/TextLeafRange.cpp	Fri May 27 04:32:40 2022 +0000
+++ b/accessible/base/TextLeafRange.cpp	Fri May 27 10:56:40 2022 +0000
@@ -1061,36 +1061,15 @@
   return attrs.forget();
 }
 
-TextLeafPoint TextLeafPoint::FindTextAttrsStart(
-    nsDirection aDirection, bool aIncludeOrigin,
-    const AccAttributes* aOriginAttrs, bool aIncludeDefaults) const {
+TextLeafPoint TextLeafPoint::FindTextAttrsStart(nsDirection aDirection,
+                                                bool aIncludeOrigin) const {
   if (IsCaret()) {
-    return ActualizeCaret().FindTextAttrsStart(aDirection, aIncludeOrigin,
-                                               aOriginAttrs, aIncludeDefaults);
+    return ActualizeCaret().FindTextAttrsStart(aDirection, aIncludeOrigin);
   }
-  // XXX Add support for spelling errors.
-  RefPtr<const AccAttributes> lastAttrs;
   const bool isRemote = mAcc->IsRemote();
-  if (isRemote) {
-    // For RemoteAccessible, leaf attrs and default attrs are cached
-    // separately. To combine them, we have to copy. Since we're not walking
-    // outside the container, we don't care about defaults. Therefore, we
-    // always just fetch the leaf attrs.
-    // We ignore aOriginAttrs because it might include defaults. Fetching leaf
-    // attrs is very cheap anyway.
-    lastAttrs = mAcc->AsRemote()->GetCachedTextAttributes();
-  } else {
-    // For LocalAccessible, we want to avoid calculating attrs more than
-    // necessary, so we want to use aOriginAttrs if provided.
-    if (aOriginAttrs) {
-      lastAttrs = aOriginAttrs;
-      // Whether we include defaults henceforth must match aOriginAttrs, which
-      // depends on aIncludeDefaults. Defaults are always calculated even if
-      // they aren't returned, so calculation cost isn't a concern.
-    } else {
-      lastAttrs = GetTextAttributesLocalAcc(aIncludeDefaults);
-    }
-  }
+  RefPtr<const AccAttributes> lastAttrs =
+      isRemote ? mAcc->AsRemote()->GetCachedTextAttributes()
+               : GetTextAttributesLocalAcc();
   if (aIncludeOrigin && aDirection == eDirNext && mOffset == 0) {
     // Even when searching forward, the only way to know whether the origin is
     // the start of a text attrs run is to compare with the previous sibling.
@@ -1104,7 +1083,7 @@
     // calculation or copying.
     RefPtr<const AccAttributes> attrs =
         isRemote ? point.mAcc->AsRemote()->GetCachedTextAttributes()
-                 : point.GetTextAttributesLocalAcc(aIncludeDefaults);
+                 : point.GetTextAttributesLocalAcc();
     if (attrs && lastAttrs && !attrs->Equal(lastAttrs)) {
       return *this;
     }
@@ -1119,7 +1098,7 @@
     }
     RefPtr<const AccAttributes> attrs =
         isRemote ? point.mAcc->AsRemote()->GetCachedTextAttributes()
-                 : point.GetTextAttributesLocalAcc(aIncludeDefaults);
+                 : point.GetTextAttributesLocalAcc();
     if (attrs && lastAttrs && !attrs->Equal(lastAttrs)) {
       // The attributes change here. If we're moving forward, we want to
       // return this point. If we're moving backward, we've now moved before